### PR TITLE
[Cocoa] Implement Alternate WebM Player Logic

### PR DESCRIPTION
**Pull request by youssefsoli**
#### 84d162afe41d5510f7b039ae9116c519db5df5cd
<pre>
[Cocoa] Implement Alternate WebM Player Logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=241949">https://bugs.webkit.org/show_bug.cgi?id=241949</a>

Reviewed by Jean-Yves Avenard.

Added the logic for the alternate WebM media player, this includes audio,
video, multi-track, fullscreen, and picture-in-picture support.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/WebMResourceClient.cpp: Added.
(WebCore::WebMResourceClient::create):
(WebCore::WebMResourceClient::WebMResourceClient):
(WebCore::WebMResourceClient::stop):
(WebCore::WebMResourceClient::dataReceived):
(WebCore::WebMResourceClient::loadFailed):
(WebCore::WebMResourceClient::loadFinished):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
(WebCore::MediaPlayerPrivateWebM::timeFudgeFactor const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::mimeTypeCache):
(WebCore::MediaPlayerPrivateWebM::supportsType):
(WebCore::MediaPlayerPrivateWebM::dataReceived):
(WebCore::MediaPlayerPrivateWebM::loadFailed):
(WebCore::MediaPlayerPrivateWebM::loadFinished):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::platformLayer const):
(WebCore::MediaPlayerPrivateWebM::play):
(WebCore::MediaPlayerPrivateWebM::pause):
(WebCore::MediaPlayerPrivateWebM::paused const):
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
(WebCore::MediaPlayerPrivateWebM::currentMediaTime const):
(WebCore::MediaPlayerPrivateWebM::seek):
(WebCore::MediaPlayerPrivateWebM::setRateDouble):
(WebCore::MediaPlayerPrivateWebM::effectiveRate const):
(WebCore::MediaPlayerPrivateWebM::setVolume):
(WebCore::MediaPlayerPrivateWebM::setMuted):
(WebCore::MediaPlayerPrivateWebM::seekable const):
(WebCore::MediaPlayerPrivateWebM::buffered const):
(WebCore::MediaPlayerPrivateWebM::setBufferedRanges):
(WebCore::MediaPlayerPrivateWebM::updateBufferedFromTrackBuffers):
(WebCore::MediaPlayerPrivateWebM::setNaturalSize):
(WebCore::MediaPlayerPrivateWebM::setHasAudio):
(WebCore::MediaPlayerPrivateWebM::setHasVideo):
(WebCore::MediaPlayerPrivateWebM::setDuration):
(WebCore::MediaPlayerPrivateWebM::setNetworkState):
(WebCore::MediaPlayerPrivateWebM::setReadyState):
(WebCore::MediaPlayerPrivateWebM::characteristicsChanged):
(WebCore::MediaPlayerPrivateWebM::createVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenFrame):
(WebCore::MediaPlayerPrivateWebM::requiresTextTrackRepresentation const):
(WebCore::MediaPlayerPrivateWebM::syncTextTrackBounds):
(WebCore::MediaPlayerPrivateWebM::setTextTrackRepresentation):
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::reenqueSamples):
(WebCore::MediaPlayerPrivateWebM::reenqueueMediaForTime):
(WebCore::MediaPlayerPrivateWebM::notifyClientWhenReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::canSetMinimumUpcomingPresentationTime const):
(WebCore::MediaPlayerPrivateWebM::setMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::clearMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::isReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::didBecomeReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::provideMediaData):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::didEncounterErrorDuringParsing):
(WebCore::MediaPlayerPrivateWebM::didProvideMediaDataForTrackId):
(WebCore::MediaPlayerPrivateWebM::append):
(WebCore::MediaPlayerPrivateWebM::abort):
(WebCore::MediaPlayerPrivateWebM::flush):
(WebCore::MediaPlayerPrivateWebM::flushIfNeeded):
(WebCore::MediaPlayerPrivateWebM::flushTrack):
(WebCore::MediaPlayerPrivateWebM::flushVideo):
(WebCore::MediaPlayerPrivateWebM::flushAudio):
(WebCore::MediaPlayerPrivateWebM::addTrackBuffer):
(WebCore::MediaPlayerPrivateWebM::ensureLayer):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::removeAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyLayer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderers):
(WebCore::MediaPlayerPrivateWebM::clearTracks):
(WebCore::MediaPlayerPrivateWebM::registerMediaEngine):
(WebCore::MediaPlayerPrivateWebM::isAvailable):

Canonical link: <a href="https://commits.webkit.org/251913@main">https://commits.webkit.org/251913@main</a>
</pre>

### CODE REVIEW

File: Source/WebCore/WebCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -14653,10 +14655,10 @@
...
-		ADDF1AD41257CD9A0003A759 /* LegacyRenderSVGPath.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = LegacyRenderSVGPath.cpp; sourceTree = "<group>"; };
-		ADDF1AD51257CD9A0003A759 /* LegacyRenderSVGPath.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = LegacyRenderSVGPath.h; sourceTree = "<group>"; };
 		ADDF1AD41257CC1A1133B644 /* RenderSVGPath.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = RenderSVGPath.cpp; sourceTree = "<group>"; };
+		ADDF1AD41257CD9A0003A759 /* LegacyRenderSVGPath.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = LegacyRenderSVGPath.cpp; sourceTree = "<group>"; };
 		ADDF1AD51257CC1A1133B644 /* RenderSVGPath.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = RenderSVGPath.h; sourceTree = "<group>"; };
+		ADDF1AD51257CD9A0003A759 /* LegacyRenderSVGPath.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = LegacyRenderSVGPath.h; sourceTree = "<group>"; };
```
Comment (youssefsoli) [inline Source/WebCore/WebCore.xcodeproj/project.pbxproj]: Unrelated sorting changes

**Comment by youssefsoli** [comment]
The style checker seems to be confused with the lambda blocks.

File: Source/WTF/wtf/PlatformEnableCocoa.h [hunk 2]
```diff
@@ -51,7 +51,7 @@
 #define ENABLE_AIRPLAY_PICKER 1
 #endif
 
-#if !defined(ENABLE_ALTERNATE_WEBM_PLAYER) && (PLATFORM(MAC) || PLATFORM(IOS) || PLATFORM(MACCATALYST))
+#if !defined(ENABLE_ALTERNATE_WEBM_PLAYER) && !PLATFORM(MACCATALYST) && !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
```
Comment (youssefsoli) [inline Source/WTF/wtf/PlatformEnableCocoa.h:54]: This was changed to address `PLATFORM(MACCATALYST)` not having `MEDIA_SOURCE` enabled.

File: Source/WebCore/platform/graphics/WebMResourceClient.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+#if ENABLE(ALTERNATE_WEBM_PLAYER)
+
+#include "ResourceError.h"
+#include "ResourceRequest.h"
+#include "ResourceResponse.h"
+
+namespace WebCore {
+
+WeakPtr<WebMResourceClient> WebMResourceClient::create(WeakPtr<WebMResourceClientParent>&& parent, PlatformMediaResourceLoader& loader, ResourceRequest&& request)
```
Comment (jyavenard) [inline Source/WebCore/platform/graphics/WebMResourceClient.cpp]: It's very unusual to return a WeakPtr or to even take a WeakPtr

It should take a ref in the constructor, up to it to only keep a WeakPtr

**Review by jyavenard — COMMENTED** [review]
LGTM other than nits above, need to fix compilation failure. Will r+ once it's green

File: Source/WebCore/platform/graphics/WebMResourceClient.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        return nullptr;
+    auto* resourcePointer = resource.get();
+    auto client = adoptRef(*new WebMResourceClient { WTFMove(parent), resource.releaseNonNull() });
+    WeakPtr result = client;
+
+    resourcePointer->setClient(WTFMove(client));
+    return result;
+}
+
+WebMResourceClient::WebMResourceClient(WeakPtr<WebMResourceClientParent>&& parent, Ref<PlatformMediaResource>&& resource)
```
Comment (jyavenard) [inline Source/WebCore/platform/graphics/WebMResourceClient.cpp]: This should take a `WebmResourceClientParent&`

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 5]
```diff
@@ -29,89 +29,220 @@
...
     MediaPlayerPrivateWebM(MediaPlayer*);
     ~MediaPlayerPrivateWebM();
-    
+
     static void registerMediaEngine(MediaEngineRegistrar);
     
+    // PlatformMediaResourceClient
+    void dataReceived(const SharedBuffer&) final;
+    void loadFailed(const ResourceError&) final;
+    void loadFinished(const FragmentedSharedBuffer&) final;
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h]: Do these need to be public?

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 6]
```diff
@@ -29,89 +29,220 @@
...
+
     void setPageIsVisible(bool) final;
-    
+
+    MediaTime timeFudgeFactor() const { return { 1, 10 }; }
     MediaTime currentMediaTime() const final;
     MediaTime durationMediaTime() const final { return m_duration; };
-    
+    MediaTime startTime() const final { return MediaTime::zeroTime(); };
+    MediaTime initialTime() const final { return MediaTime::zeroTime(); };
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:107]: Will these always be zero for WebM files? If not, it would be better to move them to the .cpp file and call `notImplemented()`

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 7]
```diff
@@ -29,89 +29,220 @@
...
+    void provideMediaData(TrackBuffer&, uint64_t);
+
+    void trackDidChangeSelected(VideoTrackPrivate&, bool);
+    void trackDidChangeEnabled(AudioTrackPrivate&, bool);
+
+    using InitializationSegment = SourceBufferParserWebM::InitializationSegment;
+    void didParseInitializationData(InitializationSegment&&);
+    void didEncounterErrorDuringParsing(int32_t);
+    void didProvideMediaDataForTrackId(Ref<MediaSampleAVFObjC>&&, uint64_t trackId, const String& mediaType);
+
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:175]: This blank line isn't needed

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h [hunk 8]
```diff
@@ -29,89 +29,220 @@
...
-    
+
     Ref<const Logger> m_logger;
     const void* m_logIdentifier;
-    
+    std::unique_ptr<VideoLayerManagerObjC> m_videoLayerManager;
+
     FloatSize m_naturalSize;
     bool m_hasAudio { false };
     bool m_hasVideo { false };
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h]: If you group all of the `bool` member variables together at the end of the object, the compiler may be able to make the object slightly smaller

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 9]
```diff
@@ -70,73 +113,956 @@
...
+{
+    ALWAYS_LOG(LOGIDENTIFIER);
+
+    auto buffer = fragmentedBuffer.makeContiguous();
+    append(buffer);
+
+    if (!m_hasVideo && !m_hasAudio) {
+        setNetworkState(MediaPlayer::NetworkState::FormatError);
+        setReadyState(MediaPlayer::ReadyState::HaveNothing);
+        return;
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: It is probably worth logging an error here

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 10]
```diff
@@ -70,73 +113,956 @@
...
+void MediaPlayerPrivateWebM::load(const String& url)
 {
     ALWAYS_LOG(LOGIDENTIFIER);
+
+    if (!m_player)
+        return;
+
+    auto mimeType = m_player->contentMIMEType();
+    if (mimeType.isEmpty() || !mimeTypeCache().contains(mimeType)) {
+        setNetworkState(MediaPlayer::NetworkState::FormatError);
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:130]: Ditto

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 11]
```diff
@@ -70,73 +113,956 @@
...
+        CachingPolicy::DisallowCaching
+    );
+    loaderOptions.destination = FetchOptions::Destination::Video;
+
+    auto loader = m_player->createResourceLoader();
+    m_resourceClient = WebMResourceClient::create(WeakPtr { *this }, *loader, WTFMove(request));
+    
+    if (!m_resourceClient) {
+        setNetworkState(MediaPlayer::NetworkState::NetworkError);
+        setReadyState(MediaPlayer::ReadyState::HaveNothing);
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:160]: Ditto

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 12]
```diff
@@ -70,73 +113,956 @@
...
+
+        FloatSize formatSize = FloatSize(PAL::CMVideoFormatDescriptionGetPresentationDimensions(formatDescription, true, true));
+        if (formatSize != m_naturalSize)
+            setNaturalSize(formatSize);
+
+        if (!m_displayLayer)
+            return;
+
+        [m_displayLayer enqueueSampleBuffer:platformSample.sample.cmSampleBuffer];
+
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: if you add an early return here you don't need the `else` below

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 13]
```diff
@@ -70,73 +113,956 @@
...
+    auto* trackBuffer = m_trackBufferMap.get(trackId);
+    if (!trackBuffer)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER, "video trackID = ", trackId, ", selected = ", selected);
+
+    if (!selected && m_enabledVideoTrackID == trackId) {
+        m_enabledVideoTrackID = -1;
+    } else if (selected) {
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: Ditto

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 14]
```diff
@@ -70,73 +113,956 @@
...
+{
+    if (trackId == m_enabledVideoTrackID) {
+        if (m_displayLayer) {
+            WeakPtr weakThis { *this };
+            [m_displayLayer requestMediaDataWhenReadyOnQueue:dispatch_get_main_queue() usingBlock:^{
+                if (weakThis)
+                    weakThis->didBecomeReadyForMoreSamples(trackId);
+            }];
+        }
+    } else if (m_audioRenderers.contains(trackId)) {
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: Ditto

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 15]
```diff
@@ -70,73 +113,956 @@
...
+
+    auto* trackBuffer = m_trackBufferMap.get(trackId);
+    if (!trackBuffer)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER, "audio trackID = ", trackId, ", enabled = ", enabled);
+
+    if (!enabled)
+        removeAudioRenderer(trackId);
+    else {
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: Ditto

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 16]
```diff
@@ -70,73 +113,956 @@
...
+
+void MediaPlayerPrivateWebM::didProvideMediaDataForTrackId(Ref<MediaSampleAVFObjC>&& originalSample, uint64_t trackId, const String& mediaType)
+{
+    UNUSED_PARAM(mediaType);
+
+    auto* trackBuffer = m_trackBufferMap.get(trackId);
+    if (!trackBuffer)
+        return;
+
+    Ref<MediaSampleAVFObjC> sample = WTFMove(originalSample);
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm]: `auto`

File: Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm [hunk 17]
```diff
@@ -70,73 +113,956 @@
...
+    reenqueSamples(m_enabledVideoTrackID);
+}
+#endif
+
+void MediaPlayerPrivateWebM::flushTrack(uint64_t trackId)
+{
+    DEBUG_LOG(LOGIDENTIFIER, trackId);
+
+    if (trackId == m_enabledVideoTrackID)
+        flushVideo();
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:906]: Early return

**Review by jyavenard — APPROVED** [review]
